x/gamm/pool-models/balancer/amm.go

```diff
@@ -0,0 +1,272 @@
+	tokenOut sdk.DecCoin, err error) {
+	tokenIn, poolAssetIn, poolAssetOut, err := p.parsePoolAssets(tokensIn, tokenOutDenom)
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	tokenAmountInAfterFee := tokenIn.Amount.ToDec().Mul(sdk.OneDec().Sub(swapFee))
+
+	poolTokenInBalance := poolAssetIn.Token.Amount.ToDec()
+	poolPostSwapInBalance := poolTokenInBalance.Add(tokenAmountInAfterFee)
+
+	// deduct swapfee on the in asset
+	// delta balanceOut is positive(tokens inside the pool decreases)
+	tokenAmountOut := solveConstantFunctionInvariant(
+		poolTokenInBalance, poolPostSwapInBalance, poolAssetIn.Weight.ToDec(),
+		poolAssetOut.Token.Amount.ToDec(), poolAssetOut.Weight.ToDec())
+	return sdk.NewDecCoinFromDec(tokenOutDenom, tokenAmountOut), nil
+}
+
+// calcInAmtGivenOut calculates token to be provided, fee added,
+// given the swapped out amount, using solveConstantFunctionInvariant
+func (p Pool) CalcInAmtGivenOut(
+	ctx sdk.Context, tokensOut sdk.Coins, tokenInDenom string, swapFee sdk.Dec) (
+	tokenIn sdk.DecCoin, err error) {
+	tokenOut, poolAssetOut, poolAssetIn, err := p.parsePoolAssets(tokensOut, tokenInDenom)
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	// delta balanceOut is positive(tokens inside the pool decreases)
+	poolTokenOutBalance := poolAssetOut.Token.Amount.ToDec()
+	poolPostSwapOutBalance := poolTokenOutBalance.Sub(tokenOut.Amount.ToDec())
+	// (x_0)(y_0) = (x_0 + in)(y_0 - out)
+	tokenAmountIn := solveConstantFunctionInvariant(
+		poolTokenOutBalance, poolPostSwapOutBalance, poolAssetOut.Weight.ToDec(),
+		poolAssetIn.Token.Amount.ToDec(), poolAssetIn.Weight.ToDec()).Neg()
+
+	// We deduct a swap fee on the input asset. The swap happens by following the invariant curve on the input * (1 - swap fee)
+	// and then the swap fee is added to the pool.
+	// Thus in order to give X amount out, we solve the invariant for the invariant input. However invariant input = (1 - swapfee) * trade input.
+	// Therefore we divide by (1 - swapfee) here
+	tokenAmountInBeforeFee := tokenAmountIn.Quo(sdk.OneDec().Sub(swapFee))
+	return sdk.NewDecCoinFromDec(tokenInDenom, tokenAmountInBeforeFee), nil
+}
+
+// ApplySwap
+func (p *Pool) ApplySwap(ctx sdk.Context, tokensIn sdk.Coins, tokensOut sdk.Coins) error {
+	// Also ensures that len(tokensIn) = 0 = len(tokensOut)
+	inPoolAsset, outPoolAsset, err := p.parsePoolAssetsCoins(tokensIn, tokensOut)
+	if err != nil {
+		return err
+	}
+	inPoolAsset.Token.Amount = inPoolAsset.Token.Amount.Add(tokensIn[0].Amount)
+	outPoolAsset.Token.Amount = outPoolAsset.Token.Amount.Sub(tokensOut[0].Amount)
+
+	return p.UpdatePoolAssetBalances(sdk.NewCoins(
+		inPoolAsset.Token,
+		outPoolAsset.Token,
+	))
+}
+
+// SpotPrice returns the spot price of the pool
+// This is the weight-adjusted balance of the tokens in the pool.
+// so spot_price = (Base_supply / Weight_base) / (Quote_supply / Weight_quote)
+func (p Pool) SpotPrice(ctx sdk.Context, baseAsset, quoteAsset string) (sdk.Dec, error) {
```

Comment:
I think its better to keep these as non-pointer receivers for guarantees around this won't mutate underlying values.

Hopefully the golang compiler eventually optimizes this, if it it deems that this could be more efficient passed by pointer. This stack vs pointer copy difference won't really affect us at the state machine level. (Avoiding a single read call to IAVL or CacheKV store saves 10x this slowdown)